### PR TITLE
feat: [AUTH-4356] M2M AuthenticateToken JWT Method

### PR DESCRIPTION
**Pull request by max-stytch**

### CODE REVIEW

File: Stytch.net.Tests/Clients/M2M.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        // Arrange
+        var client = new Stytch.net.Clients.ConsumerClient(_clientConfig);
+
+        // Act
+        // A (temporary) hack - we do not support sandbox values for M2M so we provisioned a client 
+        // that reuses the Project ID and Project Secret values we already have saved in CI 
+        var res = await client.M2M.Token(new M2MTokenRequest()
+        {
+            ClientId = _clientConfig.ProjectId,
+            ClientSecret = _clientConfig.ProjectSecret,
```
Comment (ollie-stytch) [inline Stytch.net.Tests/Clients/M2M.cs:22]: Maybe belonged in another PR but if these are required fields can we follow the pattern of having a constructor for them?

File: Stytch.net.Tests/Clients/M2M.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        // Act
+        var tokenRes = await client.M2M.Token(new M2MTokenRequest()
+        {
+            ClientId = _clientConfig.ProjectId,
+            ClientSecret = _clientConfig.ProjectSecret,
+        });
+
+        var authRes = await client.M2M.AuthenticateToken(new M2MAuthenticateTokenRequest()
+        {
+            AccessToken = tokenRes.AccessToken,
```
Comment (ollie-stytch) [inline Stytch.net.Tests/Clients/M2M.cs:46]: Same

File: Stytch.net/Clients/consumer/M2M.cs [hunk 3]
```diff
@@ -80,7 +79,56 @@ M2MTokenRequest request
...
-}
+            res.CustomClaims.Remove("scope");
 
+            if (request.RequiredScopes != null && request.RequiredScopes.Count > 0)
+            {
+                foreach (var requiredScope in request.RequiredScopes)
+                {
+                    if (!scopes.Contains(requiredScope))
+                    {
+                        throw new StytchMissingScopesException()
```
Comment (ollie-stytch) [inline Stytch.net/Clients/consumer/M2M.cs:113]: If there are multiple missing scopes we'll just return the first one?

File: Stytch.net/Clients/consumer/M2M.cs [hunk 4]
```diff
@@ -80,7 +79,56 @@ M2MTokenRequest request
...
+                Jwt = request.AccessToken,
+                ClockSkew = request.ClockSkew,
+                LifetimeValidator = request.LifetimeValidator,
+            });
+
+            var scopeClaim = (string)res.CustomClaims["scope"];
+            var scopes = new List<string>((scopeClaim).Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries));
 
-}
+            res.CustomClaims.Remove("scope");
```
Comment (ollie-stytch) [inline Stytch.net/Clients/consumer/M2M.cs:105]: Can we leave a comment here explaining what we're doing / why? It's not clear to me

File: Stytch.net/Models/Consumer/M2MClients.cs [hunk 5]
```diff
@@ -329,6 +331,48 @@ public class M2MTokenResponse
...
+
+        /// <summary>
+        /// Configures the acceptable amount of clock skew when validating timestamps.
+        /// </summary>
+        public TimeSpan ClockSkew { get; set; } = TimeSpan.FromMinutes(1);
+
+        /// <summary>
+        /// Gets or sets a delegate that will be used to validate the lifetime of the token
+        /// </summary>
+        public LifetimeValidator LifetimeValidator { get; set; }
```
Comment (ollie-stytch) [inline Stytch.net/Models/Consumer/M2MClients.cs:356]: Is this a custom lifetime validator i.e. if not specified there's some default behavior that happens?

File: Stytch.net/Utility/Utility.cs [hunk 6]
```diff
@@ -21,4 +28,76 @@ public static string BuildQueryString(Dictionary<string, string> parameters)
 
         return string.Join("&", queryString);
     }
-}
+
+    private static JsonWebKeySet _cachedJwks;
+    private static DateTime _jwksLastFetched;
+    private static TimeSpan _jwksRefreshInterval = TimeSpan.FromMinutes(15);
```
Comment (ollie-stytch) [inline Stytch.net/Utility/Utility.cs:34]: Do we have a standard for this? I could be misinterpreting but from the sounds of this message I thought we did 5 min? https://stytchio.slack.com/archives/C04LCF8GJQ6/p1729630827035289?thread_ts=1729626637.462619&cid=C04LCF8GJQ6

File: Stytch.net/Clients/consumer/M2M.cs (hunk 3 above)
Comment (max-stytch) [reply Stytch.net/Clients/consumer/M2M.cs:113]: Yeah - looking at other SDKs that's how we usually do it

File: Stytch.net/Utility/Utility.cs (hunk 6 above)
Comment (max-stytch) [reply Stytch.net/Utility/Utility.cs:34]: 5 minutes is the JWT lifetime. JWKS have a lifetime of many months, so it's kind of moot what timespan we put in. JOSE uses [ten minutes](https://github.com/panva/jose/blob/5b75eca6d0962e21e445cf932ec3b83e296e7a97/docs/jwks/remote/interfaces/RemoteJWKSetOptions.md) by default, for example.

File: Stytch.net/Models/Consumer/M2MClients.cs (hunk 5 above)
Comment (max-stytch) [reply Stytch.net/Models/Consumer/M2MClients.cs:362]: yep! There's a default validator that runs, this just allows it to be overridden for things like clock skew.

**Review by ollie-stytch — APPROVED** [review]

**Pull request merged**